File: VB/Controllers/HomeController.vb
Public Class HomeController
    Inherits Controller
    Public Function Index() As ActionResult
        Return View()
    End Function

    Public Function GridViewPartial() As ActionResult
        Return PartialView("GridViewPartial", Products)
    End Function

    Public Function NewProductPartial(ByVal product As Product) As ActionResult
        product.ID = Products.OrderByDescending(Function(p) p.ID).First().ID + 1

        Products.Add(product)

        Return PartialView("GridViewPartial", Products)
    End Function

    Public Function EditProductPartial(ByVal product As Product) As ActionResult
        Dim editProduct As Product = Products.First(Function(p) p.ID = product.ID)
        editProduct.Name = product.Name

        Return PartialView("GridViewPartial", Products)
    End Function

    Public Function DeleteProductPartial(ByVal ID As Int32) As ActionResult
        Products.Remove(Products.First(Function(p) p.ID = ID))

        Return PartialView("GridViewPartial", Products)
    End Function

    Private Property Products() As IList(Of Product)
        Get
            If HttpContext.Session("products") Is Nothing Then
                Dim productsLocal As List(Of Product) = New List(Of Product)()

                For i As Integer = 0 To 13
                    productsLocal.Add(New Product() With {.ID = i, .Name = String.Format("Product {0}", i)})
                Next i

                HttpContext.Session("products") = productsLocal
            End If

            Return CType(HttpContext.Session("products"), IList(Of Product))
        End Get
        Set(ByVal value As IList(Of Product))
            HttpContext.Session("products") = value
        End Set
    End Property
End Class

File: CS/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using Example.Models;

namespace Example.Controllers {
    public class HomeController : Controller {
        public ActionResult Index() {
            return View();
        }

        public ActionResult GridViewPartial() {
            return PartialView("GridViewPartial", Products);
        }

        public ActionResult NewProductPartial(Product product) {
            product.ID = Products.OrderByDescending(p => p.ID)
                .First()
                .ID + 1;

            Products.Add(product);

            return PartialView("GridViewPartial", Products);
        }

        public ActionResult EditProductPartial(Product product) {
            Product editProduct = Products.First<Product>(p => p.ID == product.ID);
            editProduct.Name = product.Name;

            return PartialView("GridViewPartial", Products);
        }

        public ActionResult DeleteProductPartial(Int32 ID) {
            Products.Remove(Products.First<Product>(p => p.ID == ID));

            return PartialView("GridViewPartial", Products);
        }

        private IList<Product> Products {
            get {
                if(HttpContext.Session["products"] == null) {
                    List<Product> products = new List<Product>();

                    for(int i = 0; i < 14; i++)
                        products.Add(new Product() {
                            ID = i,
                            Name = string.Format("Product {0}", i)
                        });

                    HttpContext.Session["products"] = products;
                }

                return (IList<Product>)HttpContext.Session["products"];
            }
            set {
                HttpContext.Session["products"] = value;
            }
        }
    }
}
